Improve test coverage for public models

joeydong-stripe opened the pull request:
## Summary

Improves test coverage for our public models and improves some implementations of `STPFormEncodable`

## Motivation

Better test coverage should help us iterate with more confidence.

## Testing

Verified unit tests still pass.

r? @bdorfman-stripe 
cc @bg-stripe

joeydong-stripe commented on Tests/Tests/STPBankAccountTest.m:
```diff
@@ -100,53 +117,82 @@ - (void)testBankAccountEquals {
+    XCTAssertEqualObjects(bankAccount.currency, @"usd");
+    XCTAssertEqualObjects(bankAccount.fingerprint, @"C5fW7AwE3of8bHvV");
+    XCTAssertEqualObjects(bankAccount.last4, @"6789");
+    XCTAssertNil(bankAccount.routingNumber);
```
This one's interesting. Should we map the routingNumber to the bankAccount?

bdorfman-stripe replied:
Yeah it seems like an oversight that this isn't parsed currently? Not sure if there was some intentional reason we don't parse this (although I assume not because we have a property for it). Also `routingNumber` is declared as nonnull, which is just entirely wrong. Should change that annotation and/or parse it and add it to required keys I guess.

Somewhat relatedly: It's suuuuper weird to me that STPBankAccount and STPCard are subclasses of their own Params objects. Glad we didn't do that for Sources.

What do you think @bg-stripe?

bg-stripe replied:
Yeah it is super weird – I _think_ the history there is that previously `STPCard` was used as both the params object and the returned object, and we wanted to introduce `STPCardParams` with a deprecation, rather than a breaking change.

I think the best fix here would be to make `STPBankAccount` not a subclass of `STPBankAccountParams` and do the same with `CardParams` – maybe we should save that for a separate PR though.

joeydong-stripe replied:
Cool, I'll at least map the `routingNumber`. I think overall `STPBankAccount` has a few nullability bugs that we can fix (cause public api changes) but I'm guessing the object isn't used very often so we don't hear much about it.

Are we ever using the fact that `STPBankAccount` is a subclass of `STPBankAccountParams` to allow people to retrieve/modify/resubmit a bank account easily? Maybe that's too magical.

joeydong-stripe commented:
Sorry for the commit churn. Also, I can break this PR up if this is too much at once.

bdorfman-stripe commented on Stripe/STPBankAccountParams+Private.h:
```diff
@@ -15,6 +15,7 @@ NS_ASSUME_NONNULL_BEGIN
 + (STPBankAccountHolderType)accountHolderTypeFromString:(NSString *)string;
 + (NSString *)stringFromAccountHolderType:(STPBankAccountHolderType)accountHolderType;
 
+- (instancetype)init;
```
Does this need to be here? Don't all NSObject's inherit init anyway?

joeydong-stripe replied:
I think I was getting at making sure we have an explicit designated initializer for everything and seeing if that provides any compiler safety but didn't finish it. I'll remove this for now and restart the investigation later.

bdorfman-stripe commented on Tests/Tests/STPBankAccountTest.m:
```diff
@@ -8,21 +8,20 @@
 - (void)setUp {
     [super setUp];
-    _bankAccount = [[STPBankAccount alloc] init];
+    // Put setup code here. This method is called before the invocation of each test method in the class.
```
If there is no setup needed, you should just delete the method.

joeydong-stripe replied:
👍

bdorfman-stripe commented on Tests/Tests/STPBankAccountTest.m:
```diff
@@ -100,53 +117,82 @@ - (void)testBankAccountEquals {
 
 #pragma mark - STPAPIResponseDecodable Tests
 
 - (NSDictionary *)completeAttributeDictionary {
```
Test json like this should probably be moved to STPFixtures? And/or an actual json file.

joeydong-stripe replied:
👍

bdorfman-stripe commented on Tests/Tests/STPCardTest.m:
```diff
@@ -13,15 +13,13 @@
 - (void)setUp {
     [super setUp];
-    _card = [[STPCard alloc] init];
+    // Put setup code here. This method is called before the invocation of each test method in the class.
```
Same as other setup comment.

joeydong-stripe replied:
👍

bdorfman-stripe commented on Tests/Tests/STPCardTest.m:
```diff
@@ -151,38 +149,86 @@ - (void)testStringFromFunding {
 - (void)testLast4ReturnsCardNumberLast4WhenNotSet {
-    self.card.number = @"4242424242424242";
-    XCTAssertEqualObjects(self.card.last4, @"4242", @"last4 correctly returns the last 4 digits of the card number");
+    STPCard *card = [[STPCard alloc] init];
```
can just do `[STPCard new]` instead of alloc/init? Or is that a style guide holy war? :)

bg-stripe replied:
:P +1 for `new` in the style guide

joeydong-stripe replied:
I added this as one of our style guide discussion points 😉

bdorfman-stripe commented on Tests/Tests/STPCardTest.m:
```diff
@@ -191,66 +237,118 @@ - (void)testCardEquals {
-             @"dynamic_last4": @"5678",
-             @"brand": @"MasterCard",
-             @"country": @"Japan",
+             @"id": @"card_1AVRojEOD54MuFwSxr93QJSx",
```
Where do these ids actually come from? I have different ones on my screen. Are we sure these aren't "real" tokens that are populated from the logged in Stripe account or something?

Also again I think all fake json stuff should be through STPFixtures now.

bg-stripe replied:
(I usually use dummy IDs like `card_123` in fixtures)

joeydong-stripe replied:
haha yeah looks like it's one of my dummy test cards. i'll figure something out.

joeydong-stripe commented on Tests/Tests/STPCardTest.m:
```diff
@@ -151,106 +142,176 @@ - (void)testStringFromFunding {
+    XCTAssertEqualObjects(card.name, @"Jane Austen");
+
+    XCTAssertNotEqual(card.allResponseFields, response);
+    XCTAssertEqualObjects(card.allResponseFields, [response stp_dictionaryByRemovingNullsValidatingRequiredFields:@[]]);
```
I'll come back to fixing this `stp_dictionaryByRemovingNullsValidatingRequiredFields` usage in another PR. The category is missing test coverage and we should probably deeply remove `[NSNull null]`s.

joeydong-stripe commented:
wdyt @bdorfman-stripe ?

joeydong-stripe commented:
Just moved all of the private header definitions I added that were specific to tests in 0a89c95 because it looks like that's where our style is headed.

cc @bg-stripe
